### PR TITLE
More annotations

### PR DESCRIPTION
**Pull request by crusaderky**
A little incremental addition to type annotations. By no means complete, but it should be ready for merge in its own right nonetheless.

### CODE REVIEW

File: xarray/core/dataset.py [hunk 1]
```diff
@@ -1205,24 +1209,28 @@ def indexes(self) -> 'Mapping[Any, pd.Index]':
...
-    def data_vars(self):
+    def data_vars(self) -> DataVariables:
         """Dictionary of DataArray objects corresponding to data variables
         """
         return DataVariables(self)
 
-    def set_coords(self, names, inplace=None):
+    def set_coords(
+        self,
+        names: 'Union[Hashable, Iterable[Hashable]]',
```
Comment (crusaderky) [inline xarray/core/dataset.py]: It would have been nicer to use Collection[Hashable], but Collection requires Python 3.6

Comment (crusaderky) [reply xarray/core/dataset.py]: Note: Sequence is NOT ok, because sets are not sequences

File: xarray/core/dataset.py [hunk 2]
```diff
@@ -1241,19 +1249,26 @@ def set_coords(self, names, inplace=None):
         # nb. check in self._variables, not self.data_vars to insure that the
         # operation is idempotent
         inplace = _check_inplace(inplace)
-        if isinstance(names, str):
+        if isinstance(names, str) or not isinstance(names, Iterable):
             names = [names]
+        else:
+            names = list(names)
```
Comment (crusaderky) [inline xarray/core/dataset.py]: This forced casting could have been avoided with Collection :(

File: xarray/core/dataset.py [hunk 3]
```diff
@@ -2974,7 +2992,11 @@ def unstack(self, dim=None):
             result = result._unstack_once(dim)
         return result
 
-    def update(self, other, inplace=None):
+    def update(
+        self,
+        other: 'Union[Dataset, Mapping[Hashable, DataArray]]',
```
Comment (crusaderky) [inline xarray/core/dataset.py]: unsure about this - are there more options available for 'other'?

Comment (shoyer) [reply xarray/core/dataset.py]: This goes through the `merge` machinery (same as `Dataset.__init__`), so in principle I think you can put `Variable` objects or tuples of Variable arguments in the mapping values, too.

File: xarray/core/common.py [hunk 4]
```diff
@@ -161,13 +161,15 @@ class AttrAccessMixin:
     _initialized = False
 
     @property
-    def _attr_sources(self):
-        """List of places to look-up items for attribute-style access"""
+    def _attr_sources(self) -> List[Mapping[Hashable, Any]]:
```
Comment (max-sixty) [inline xarray/core/common.py]: Is `Mapping[Hashable, Any]` identical to `Mapping`?

Comment (crusaderky) [reply xarray/core/common.py]: There is no requirement for the keys of a Mapping to be hashable. This is relevant for the various setter methods Dataset.attrs which cast any Mapping to an OrderedDict.

**Comment by shoyer** [comment]
thanks @crusaderky !

**Pull request merged**